### ISSUE TITLE
Command to temporarily have text pass \ifpages test

### ISSUE THREAD

**Issue by moewew** [plk/biblatex#918]
This came up in https://tex.stackexchange.com/q/506360/35864

In some cases it might be useful to have a command to allow postnote text to pass the `\ifpages` test. This would be useful for some `loccit` operations or for `citepages` features.

Simple example
```latex
\documentclass[british]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{babel}
\usepackage{csquotes}

\usepackage[style=authoryear, backend=biber]{biblatex}

\DeclareFieldFormat{postnote}{%
  \ifpages{#1}
    {pages}
    {no pages}
  #1}

\addbibresource{biblatex-examples.bib}


\begin{document}
\cite[380]{sigfridsson}

\cite[381 and more]{sigfridsson}

\cite[381]{sigfridsson}

\printbibliography
\end{document}
```

Example showing `loccit` trackers and `citepages` options in action
```latex
\documentclass[british]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{babel}
\usepackage{csquotes}

\usepackage[style=verbose-ibid, ibidpage=true, citepages=separate, backend=biber]{biblatex}

\addbibresource{biblatex-examples.bib}


\begin{document}
\cite[380]{sigfridsson}

\cite[380]{sigfridsson}

\citereset

\cite[381 and more]{sigfridsson}

\cite[381 and more]{sigfridsson}
\end{document}
```

**Comment by moewew** [comment]
Since https://github.com/plk/biblatex/commit/235ffcee878d4c78f63d489730f1d165306863f0 this is possible as follows
```latex
\documentclass[british]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{babel}
\usepackage{csquotes}

\usepackage[style=verbose-ibid, ibidpage=true, citepages=separate, backend=biber]{biblatex}

\addbibresource{biblatex-examples.bib}

\makeatletter
\let\passifpages\@firstofone
\PagesCheckSetup{\let\passifpages\@gobble}
\makeatother

\begin{document}
\cite[380]{sigfridsson}

\cite[380]{sigfridsson}

\citereset

\cite[\passifpages{\pno~381 and more}]{sigfridsson}

\cite[\passifpages{\pno~381 and more}]{sigfridsson}
\end{document}
```

The definitions of `\NumsCheckSetup` and `\PagesCheckSetup` are trivial
```latex
\newrobustcmd*{\NumsCheckSetup}{\appto\blx@hook@ifnums}
\newrobustcmd*{\PagesCheckSetup}{\appto\blx@hook@ifpages}
```

**Comment by moewew** [comment]
@dcpurton Something like the following should work for your answer in https://tex.stackexchange.com/q/506360/35864

```latex
\documentclass{article}
\usepackage[ngerman]{babel}

\usepackage[german=quotes]{csquotes}
\usepackage[style=sbl, citepages=separate]{biblatex}

\makeatletter
\providerobustcmd*{\NumsCheckSetup}{\appto\blx@hook@ifnums}
\providerobustcmd*{\PagesCheckSetup}{\appto\blx@hook@ifpages}

\let\passifpages\@firstofone
\PagesCheckSetup{\let\passifpages\@gobble}
\makeatother

\renewbibmacro*{pages}{%
  \iffieldundef{pages}
    {\ifbool{bbx@inset}
       {}
       {\printfield{postnote}%
        \global\booltrue{suppresspostnote}}}
    {\printfield{pages}%
     \ifboolexpr{
       test {\iffieldundef{postnote}}
       or
       bool {bbx@inset}
     }
       {}
       {\setunit{\addspace}%
         \printtext[parens]{%
          \iffieldpages{postnote}
            {\bibstring{thiscite}%
             \setunit{\addspace}}
            {}%
          \printfield{postnote}}}}}

\begin{filecontents*}{\jobname.bib}
@book{Pritchard:1969,
  editor    = {Pritchard, James B.},
  title     = {Ancient Near Eastern Texts Relating to the Old Testament},
  edition   = {3},
  location  = {Princeton},
  publisher = {Princeton University Press},
  date      = {1969},
}
@article{Kriger.1907,
  author       = {Kriger, John},
  date         = {1907},
  title        = {Forum Romanun},
  shorttitle   = {Forum},
  pages        = {11--14},
  volume       = {6},
  journaltitle = {Le Muséon},
  shortjournal = {LM},
}
@review{Wellhausen.1906,
  author={Wellhausen, Julius},
  revdauthor   = {Ban, Paul},
  revdtitle    = {select, Paris/ Leipzig 1905},
  journaltitle = {Le Museon},
  shortjournal = {LM},
  volume       = {8,2},
  date         = {1906},
  pages        = {16--18},
}
\end{filecontents*}
\addbibresource{\jobname.bib}

\begin{document}
\printbiblist{abbreviations}

Die Debatten sind nach Meinung von Wellhausen \enquote{ein AUsdruck zeitgenössischer Debatten“}\autocite[164]{Wellhausen.1906}.
An anderer Stelle wird er sogar noch deutlicher.\autocite[Siehe auch][]{Pritchard:1969}
Grer\footnote{Vgl. \cite[\passifpages{11 [dt.]; 12 [engl.]}]{Kriger.1907}.}
Grer\footnote{\cite[11.12.13]{Kriger.1907}.}

sdf\footnote{Vgl. \cite[\passifpages{11 [dt.]; 12 [engl.]}]{Kriger.1907}.}
adfist aufgelistet

\printbibliography
\end{document}
```

**Comment by moewew** [comment]
Another use case is in https://tex.stackexchange.com/q/531646/35864.

MWE for use with `biblatex` 3.15 and above
```latex
\documentclass[british]{article}
\usepackage{csquotes}
\usepackage{babel}
\usepackage[style=verbose, citepages=omit]{biblatex}
\usepackage[colorlinks]{hyperref}

\makeatletter
\newcommand*{\hidefromifpages}[1]{#1}
\PagesCheckSetup{\let\hidefromifpages\@gobble}
\makeatother

\addbibresource{biblatex-examples.bib}

\begin{document}
\null\vfill\noindent
text.\footcite[\pnfmt{380--381}\hidefromifpages{, textual comment}]{sigfridsson}
\printbibliography
\end{document}
```

Messed up the issue link in https://github.com/plk/biblatex/commit/e066a75395d13943324973db6ad0c7b0732c2d5a, see also https://github.com/plk/biblatex/commit/263d9486558bd4e9cd1040f50c746fa589b6281c.

**Comment by zaeph** [comment]
I've got a related issue.

I was looking for a way to make `ibidpage=true` work nicely with `emphasis added` in the postnote, i.e., by excluding it from the `same-page` cond-check.  I thought I'd found the grail with #918, but it's not quite _it_.

With the MWE you've proposed, I expected this to work:
```tex
\footcite[\passifpages{\pnfmt{319}}, emphasis added]{declercq2018}
\footcite[\passifpages{\pnfmt{319}}, emphasis added]{declercq2018}
```
This is because I expected `\passifpages` to only act on the page-section of the postnote, but it doesn't look like there's any way to do it with the current code.

[Someone on SX managed to cobble up something together to structure the postnote](https://tex.stackexchange.com/a/95205), but I think we could come up with a better implementation.

How about adding another optional argument to the `cite`-commands?
```tex
\cite[post]{key}
\cite[pre][post]{key}
\cite[pre][post-page][post-text]{key}
```
I don't think this would change the default behaviour of the `cite`-commands, but I'm not familiar enough with the codebase to know for sure.

Another implementation would be to patch `\blx@loccit@stricttracker` to have a way to match location structures:
https://github.com/plk/biblatex/blob/abfeeb95352a77e69a700b9e3daa3c95c47966d5/tex/latex/biblatex/biblatex.sty#L11161-L11173
An example would be to have it parse `[{<location-structure>}…]`, format it with `\pnfmt` when `<location-structure>` is numerical, and check the equality of `<location-structure>` when `ibidpage=true`.

**Comment by moewew** [comment]
I'll have to dash, so can't look at this in more detail now, but why don't you `\passifpages` everything?

```latex
\documentclass[british]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{babel}
\usepackage{csquotes}

\usepackage[style=verbose-ibid, ibidpage=true, citepages=separate, backend=biber]{biblatex}
\addbibresource{biblatex-examples.bib}

\makeatletter
\let\passifpages\@firstofone
\PagesCheckSetup{\let\passifpages\@gobble}
\makeatother

\begin{document}
Yes\footcite[\passifpages{\pnfmt{319}, emphasis added}]{sigfridsson}
Yes\footcite[\passifpages{\pnfmt{319}, emphasis added}]{sigfridsson}

No\footcite[\passifpages{\pnfmt{319}}, emphasis added]{nussbaum}
No\footcite[\passifpages{\pnfmt{319}}, emphasis added]{nussbaum}

\printbibliography
\end{document}
```

**Comment by zaeph** [comment]
> I'll have to dash, so can't look at this in more detail now, but why don't you `\passifpages` everything?

Because, even though it's on the same page, I'm referring to another quote from that page, and `emphasis added` is warranted because I've also added emphasis to that second quote.

Here's the desired output:
```
1. Declercq 2018, p. 319.
2. Ibid., emphasis added.
3. Ibid., emphasis added.
```

To be clear, I think that #918 is a move in a good direction, but I think that what I've suggested above might make for a more robust implementation.  I'll readily admit that it would have a bigger footprint than your `\passifpages`.  There might be even be a way to only suppress the location-structure and keep the rest of the postnote with that implementation.

**Comment by moewew** [comment]
Ah, I see. That is quite a bit more complicated than what we have here so far, I suggest you open a new issue for this (ideally with a short example document we can play around with), so it doesn't get lost here (at the bottom of what is essentially a resolved issue).

I haven't given any of this proper thought, but as far as I can see this would essentially require a possibility to completely decouple a page range in the post note from additional postnote text. That appears to be quite tricky if we don't want to introduce an arbitrary new interface like suggestion 2. A new `post-postnote` argument to `\...cite` looks like the most attractive option to me from a conceptual point of view, but I don't think I'd ever dare going for it: It would be quite the departure from what has essentially be a stable interface for a decade in `biblatex` and for much longer before that in similar form in `natbib`.

**Comment by zaeph** [comment]
> I suggest you open a new issue for this (ideally with a short example document we can play around with), so it doesn't get lost here (at the bottom of what is essentially a resolved issue).

Got it.  I'm wrapping up my thesis this week, but I'll file an issue next week.

> It would be quite the departure from what has essentially be a stable interface for a decade in `biblatex` and for much longer before that in similar form in `natbib`.

Tell you what: I'll try my hand at the implementation to wrap my hand around the codebase, and then we can reassess whether to merge it in `master`.  Otherwise, we can just push it to CTAN as an extra package.

**Comment by zaeph** [comment]
I've got my hands a bit full right now with other projects, so I'm going to have to put this one on stand-by.  I'll try to get back to it later this month.

**Comment by moewew** [comment]
The points discussed in my initial points have been addressed in `biblatex` 3.15 (which is available now on TeX live and MikTeX).

If you are still interested in further discussions, @zaeph, please open a new issue. I'm closing this for now.
